### PR TITLE
Implement the european notation

### LINKED ISSUE #401: European notation support
Hi! Thanks for your work.

I was wondering if you were planning to introduce support to the European notation (DO, RE, MI, FA, SOL, LA, SI).
To my knowledge this isn't yet a feature of this library.

Do you think this is feasible?

I could make a pull request and implement it myself if needed, but I think I'm going to require some tips on which files to modify.

### PR DESCRIPTION
**Pull request by SamuMazzi**
Actually in the code the european notation is referred as `solfege` (I just didn't feel right to call it "european" because of the "german" notation, and online they also called it "solfege" so I went for this direction but if it's a problem it can be easily changed) and it's treated like other ChordType notations, like `symbol`/`numeric`/`numeral`

Also I implemented some tests and modified the existing ones, but some tests are left, but that's because I wanted to commit this to understand if my work's going in the right direction or if I should change something (or everything :')

Any feedback is really welcome, and thanks for your job!

Resolves #401

### CODE REVIEW

**Comment by SamuMazzi** [comment]
I cannot really fix these eslint issues in this weekend but if you give me a feedback I'll finish it next week!

**Comment by martijnversluis** [comment]
@SamuMazzi Thanks for this great pull request, you really put the effort in it!

Coming week I'll take the time to go through it and check the CodeClimate issues. Will get back to you 👍

**Comment by martijnversluis** [comment]
@SamuMazzi I resolved the CodeClimate issues and ran the build. Could you check the ESlint issues?

**Comment by SamuMazzi** [comment]
Solved it! I've also noticed that I've mistakenly changed some dependencies.. I can revert those changes or you can keep them (everything was obviously still working)

File: test/formatter/text_formatter.test.ts [hunk 1]
```diff
@@ -32,7 +32,37 @@ Grid line
 Tab 1
 Tab line`.substring(1);
 
-    expect(new TextFormatter().format(song)).toEqual(expectedChordSheet);
+    expect(new TextFormatter().format(exampleSongSymbol)).toEqual(expectedChordSheet);
+  });
+
+  it('formats a solfege song to a text chord sheet correctly', () => {
+    // FIXME: Fix "wis dom" spaced
```
[the remark this answers is not in the document]
Comment (martijnversluis) [reply test/formatter/text_formatter.test.ts]: I think the spacing is correct, else the chords are displayed as `SolLa`

File: test/helpers/render_chord.test.ts [hunk 2]
```diff
@@ -5,19 +5,19 @@ import { eachTestCase } from '../utilities';
...
-       1 |         |      |         |                  |           | "Em7"   |
-       2 |         |      |         |                  | "F"       | "Em7"   |
-       3 |         |      |         | "A"              |           | "Em7"   |
-       4 |         |      |         | "A"              | "F"       | "Em7"   |
-       5 |         |      | "Bb"    |                  | "F"       | "Em7"   |
-       6 |         |      | "Bb"    | "A"              | "F"       | "Em7"   |
-       7 |         | 3    |         |                  |           | "Dbm7"  |
-       8 |         | 3    |         |                  | "F"       | "Dbm7"  |
-       9 |         | 3    |         | "A"              |           | "Dbm7"  |
+        1 |         |      |         |                  |           | "Em7"   |
```
Comment (martijnversluis) [inline test/helpers/render_chord.test.ts:12]: I think these whitespace changes should be excluded

File: test/helpers/render_chord.test.ts [hunk 3]
```diff
@@ -53,24 +53,76 @@ describe('renderChord helper', () => {
...
-       1 | "Gm"        |         | "symbol"   | "Gm"    |
-       2 | "Gm"        | "Bb"    | "symbol"   | "Gm"    |
-       s | "Gm"        | "Bb"    | "number"   | "6"     |
-       s | "Gm"        | "Bb"    | "numeral"  | "vi"    |
-       5 | "IV"        |         | "numeral"  | "IV"    |
-       6 | "IV"        |         | "number"   | "4"     |
-       7 | "IV"        | "Bb"    | "symbol"   | "Eb"    |
-       8 | "IV"        | "Bb"    | "numeral"  | "IV"    | 
-       9 | "IV"        | "Bb"    | "number"   | "4"     |
+        1 | "Gm"        |         | "symbol"   | "Gm"    |
```
Comment (martijnversluis) [inline test/helpers/render_chord.test.ts:108]: I think these whitespace changes should be excluded

File: src/key_config.ts [hunk 4]
```diff
@@ -1,69 +1,136 @@
...
+      1: 'Bm', 3: 'Am', 6: 'F#m', 8: 'Em',
+    },
+    'C#m': {
+      2: 'Bm', 4: 'Am', 7: 'F#m', 9: 'Em',
+    },
+    'Dm': { 3: 'Bm', 5: 'Am', 8: 'F#m' },
+    'Ebm': { 1: 'Dm', 4: 'Bm', 6: 'Am' },
+    'D#m': { 1: 'Dm', 4: 'Bm', 6: 'Am' },
+  },  
+}
```
Comment (martijnversluis) [inline src/key_config.ts]: Trailing linefeed is missing here

**Comment by martijnversluis** [comment]
One thing I'm a bit uncertain about (I've never used the solfege system), is whether the solfege keys are actual "absolute" keys, so if solfege is similar to chord symbols. I always believed solfege is more like numbers of numerals, where it's just a relative note in the current key. 

I hope maybe @isaiahdahl knows?

**Comment by isaiahdahl** [comment]
@martijnversluis its actually both! There is "Movable Do" and "Fixed Do"

	•	Movable Do: Predominantly used in English-speaking countries. 
	•	Fixed Do: More common in European countries like Italy, France, and Spain, and also in many Latin American countries. In fixed do "C" = "do" "E" = "re" 

From my reading one isn't technically more "popular" then the other across the board. 

We'll probably want to consider a way to leave room to support both types. Like have "fixed" or "moveable" in the configuration somewhere. Maybe they need separate implementations completely with naming to differentiate them. Like movableDo/ fixedDo instead of solfege.

Just my thoughts!

File: src/key_config.ts (hunk 4 above)
Comment (SamuMazzi) [reply src/key_config.ts]: You're actually referring to an old commit.. in the newers it's already fixed.

**Comment by SamuMazzi** [comment]
I've never known about the movable Do actually, thanks for the info @isaiahdahl!

As I said at the beginning, I just chose the name that seemed more popular and "explicative", but at this point tell me how you want to name it.
And for the movableDo it seems really similar to number/numerals so maybe it's not gonna need a complete different implementation

**Comment by isaiahdahl** [comment]
After a bit more thought here are some follow-up opinions.

I think the name `solfege` as a name for the chord type is right, it's just a matter of finding a way to support both moveable Do and fixed Do in a clean way. So meaning we should think of `solfege` as a chord type, regardless of whether it was moveable Do or not.

I don't think it makes sense to have two different chord "types" for movable and fixed, because fundamentally they are the same parsable chord at that level. It's only when you start changing keys and transposing that the movable/fixed comes into play. 

I think if we had a `toSolfege(key = null)` function that had logic built into it to read from the config we could handle the differences that way.   

There should be a way for someone to take a chart that was written with symbols, turn it into a solfege chart, then to numbers, then back to symbols and have it spit the same chart out. And have it possible to do that in a mode where solfege is movable, and where it's fixed. 

Another thought after more general code reviewing with this new type is maybe the test directory for chords should be structured differently to make this a little more clear. What if instead of ./test/chord_symbol and ./test/chord_solfege directories, we moved what's in the chord_symbol test directory into the ./test/chord directory and then had sub folders in there for each type:

```
./test
└── chord
    ├── solfege
    ├── symbol
    └── numerals

```

What lead me to think this is seeing this test: Which doesn't really make sense to me why we would need/want this in both the chord_symbol dir and chord_solfege dir

./test/is_numeric.test.ts
```ts
import { Chord } from '../../src';

describe('Chord', () => {
  describe('chord solfege', () => {
    describe('for a pure numeric chord', () => {
      it('returns true for a numeric chord', () => {
        expect(Chord.parse('1/3')?.isNumeric()).toBe(true);
      });
    });
  });
});

```

Then we would make sure that there is a test in the `symbol` directory for to_chord_solfege_fixed.test.ts that has some test for turning chords of type 'numeral' | 'symbol' | 'numbers' into a solfege type chord that was handling the fixed/moveable stuff as well. 

Thoughts @martijnversluis

**Comment by SamuMazzi** [comment]
Just to clarify I had to copy-paste some tests, like in this case, but I totally agree on the new format, removing these kind of tests that don't really make sense for me.
Regarding the new moveable Do logic, if we decide to implement it in this PR, I'll definetely need someone to help me to understand better how this works in theory

But I'll wait for what you have to say @martijnversluis

**Comment by martijnversluis** [comment]
> Solved it! I've also noticed that I've mistakenly changed some dependencies.. I can revert those changes or you can keep them (everything was obviously still working)

Missed this comment. Yeah, I like dependency updates in separate PRs, so if it's no trouble please revert those.

**Comment by martijnversluis** [comment]
I appreciate this thorough discussion!

I agree with @isaiahdahl that movable and fixed solfege should be the same implementation, which mainly lives in the chord parsing logic. 

- parsing fixed solfege should probably work like parsing chord symbols
- parsing movable solfege should work like parsing numbers/numerals.

I also agree with the restructuring of tests.

_However_: as far as I can see does this PR deliver a working implementation of fixed solfege. What about merging this part, and then restructure the tests and try adding support for movable solfege in two seperate PRs?

**Comment by SamuMazzi** [comment]
I definitely agree with you! I reverted those packages changes so I'd say that this PR is okay now.

And for the movable solfege I could take care of that, but still it's not really clear in my mind how that works, but maybe it's better to open an issue and talk there about that!

**Comment by SamuMazzi** [comment]
@isaiahdahl @martijnversluis I don't want to bother you but are there any news on this? I'd need this version to be merged with the main branch for a project of mine, but if there are other things to change let's talk about it
Thank you :)

**Comment by martijnversluis** [comment]
Sorry to have kept you waiting @SamuMazzi! Merged this into `master` just now.

**Pull request merged**